rwatch/ui/animation/property_animation.c

```diff
@@ -0,0 +1,172 @@
+    
+    property_animation->impl.accessors.setter.uint32(uint32, (int)(((to - from) / 100.0)*progress));
+}
+
+void property_animation_update_gcolor8(PropertyAnimation * property_animation, const uint32_t distance_normalized)
+{
+    int progress = (distance_normalized * 100) / ANIMATION_NORMALIZED_MAX;
+    
+    GColor8 *gcolor = (GColor8 *) property_animation->subject;
+    
+    GColor8 *from = &property_animation->values.from.gcolor;
+    GColor8 *to = &property_animation->values.to.gcolor;
+    
+    GColor8 new_gcolor = *from;
+    new_gcolor.r = from->r + (((to->r - from->r) / 100.0)*progress);
+    new_gcolor.g = from->g + (((to->g - from->g) / 100.0)*progress);
+    new_gcolor.b = from->b + (((to->g - from->b) / 100.0)*progress);
+    new_gcolor.a = from->a + (((to->a - from->a) / 100.0)*progress);
+    if (property_animation->impl.accessors.setter.gcolor)
+        property_animation->impl.accessors.setter.gcolor(gcolor, new_gcolor);
+}
+
+PropertyAnimation * property_animation_create_layer_frame(struct Layer * layer, GRect * from_frame, GRect * to_frame)
+{
+    const PropertyAnimationImplementation implementation = {
+        .base = {
+            .update = (AnimationUpdateImplementation) property_animation_update_grect,
+        },
+        .accessors = {
+            .setter = { .grect = (GRectSetter) layer_set_frame },
+            .getter = { .grect = (GRectGetter) layer_get_frame },
+        },
+    };
+    
+    PropertyAnimation *property_anim = property_animation_create(&implementation, layer, from_frame, to_frame);
+    return property_anim;
+}
+
+PropertyAnimation * property_animation_create_bounds_origin(struct Layer * layer, GPoint * from, GPoint * to)
+{
+    const PropertyAnimationImplementation implementation = {
+        .base = {
+            .update = (AnimationUpdateImplementation) property_animation_update_gpoint,
+        },
+        .accessors = {
+            .setter = { .gpoint = layer_get_bounds },
+            .getter = { .gpoint = layer_get_bounds },
+        },
+    };
+    
+    PropertyAnimation *property_anim = property_animation_create(&implementation, layer, from, to);
+    
+    return property_anim;
+}
+
+PropertyAnimation * property_animation_create(const PropertyAnimationImplementation * implementation, void * subject, void * from_value, void * to_value)
+{
+    SYS_LOG("property_animation", APP_LOG_LEVEL_INFO, "property_animation_create");
+    
+    PropertyAnimation *property_animation = app_calloc(1, sizeof(PropertyAnimation));
+    Animation *animation = animation_create();
+    
+    property_animation->animation = animation;
+    property_animation->animation->property_anim = property_animation;
+    property_animation->animation->impl = implementation->base;
+    property_animation->impl = *implementation;
+    property_animation->subject = subject;
+    
+    // Check the type of the values
+    if (((GRect *) from_value)->size.w != NULL)
```

Comment:
I think we can just store pointers to from_value and to_value, rather than having to guess the type

Reply:
They'll probably be destroyed when the function the animation is created in returns (maybe I'm thinking of that wrong). It depends on how the developer wrote their part.